src/libraries/System.Net.Quic/src/System/Net/Quic/QuicListener.cs

```diff
@@ -166,15 +176,19 @@ public async ValueTask<QuicConnection> AcceptConnectionAsync(CancellationToken c
         GCHandle keepObject = GCHandle.Alloc(this);
         try
         {
-            PendingConnection pendingConnection = await _acceptQueue.Reader.ReadAsync(cancellationToken).ConfigureAwait(false);
-            await using (pendingConnection.ConfigureAwait(false))
+            object item = await _acceptQueue.Reader.ReadAsync(cancellationToken).ConfigureAwait(false);
+            Interlocked.Increment(ref _pendingConnectionsCapacity);
+
+            if (item is QuicConnection connection)
             {
-                return await pendingConnection.FinishHandshakeAsync(cancellationToken).ConfigureAwait(false);
+                return connection;
             }
+            ExceptionDispatchInfo.Throw((Exception)item);
```

Comment:
Do we really want to do it? For TCP the accept does not return any failed connections. While this may be handy for debugging, I'm worried about DoS when somebody can flood init attempt and we would go to throwing/catching game making it way more expensive IMHO. I would suggest log and silent discard ASAP.

cc: @JamesNK @Tratcher for any more thoughts.

Reply:
Kestrel stops listening to `QuicListener.AcceptConnectionAsync` if the methods throws:
1. `QuickException` with `QuicError.OperationAborted`
2. `ObjectDisposedException`
3. `OperationCanceledException` triggered by canceling the passed-in token.

Anything else is logged, and Kestrel calls accept again.

Reply:
This has been discussed and changed and is nothing new in this PR, see https://github.com/dotnet/runtime/issues/72160#issuecomment-1184127586, https://github.com/dotnet/runtime/issues/72161#issuecomment-1184138201 and the PR https://github.com/dotnet/runtime/pull/72390. Silently discarding errors was causing troubles diagnosing some problems.

TCP may not give you failed accept attempts, but errors from establishing SSL on top of that would get propagated, wouldn't they? And this is more like that IMHO. Note that this exception get here only intentionally from `StartConnectionHandshake` and these are the cases:
- options validation failed, i.e. user provided wrong options on the callback, including errors from preparing `MsQuicConfiguration`, this is a way to let them know
- TLS handshake failed for any reason, i.e. an error coming from MsQuic
- handshake timeout expired (10 seconds), i.e our own imposed as MsQuic would never kill that connection in this state: https://github.com/microsoft/msquic/discussions/2705

On top of that, we are actively refusing any connection attempts over `ListenBacklog` count, without even starting the handshake. Which is something we didn't do in 6.0 where we discarded that connection only after we fully established it. 

We can fine tune the cases when we propagate the errors, but I wouldn't discard them all, that's something that's been intentionally fixed.

Reply:
I'm not convinced about the arguments, but I can live with the outcome. We should also update https://learn.microsoft.com/en-us/dotnet/api/system.net.quic.quiclistener.acceptconnectionasync?view=net-7.0 and perhaps be specific what exceptions `AcceptConnectionAsync` can throw. 

For background, AAD has benchmark to measure how expensive it is to handle failed connections/requests. And I think the DoS scenarios are real e.g. cost of error handling. But it can perhaps still be small comparing to cost of the crypto.